architecture/2025-01-serverless-feature-spec.md

```diff
@@ -0,0 +1,534 @@
+
+> Note: Currently it's one entry per `extension` for Kubernetes, but as a part of implementation we should determine if this needs to be further nested into one `extension` per platform and within each platform's `extension` can have multiple entries.
+
+### Step 2: Define a Radius Application for serverless
+
+The user defines a new Radius Application by creating a new Application definition file (e.g. `app.bicep`) and uses this same application definition to deploy to an Environment with either Kubernetes or serverless as its underlying compute platform. Via the `extension` property in the Application definition, the user can optionally configure custom settings that are specific to the container runtime platform and get applied only when deploying to that platform. For any `extension` properties that are not applicable to the targeted deployment container platform, they will be ignored.
+
+```diff
+@description('The environment ID of your Radius Application. Set automatically by the rad CLI.')
+param environment string
+
+resource app 'Applications.Core/applications@2023-10-01-preview' = {
+  name: 'myapp'
+  properties: {
+    environment: environment
++   extensions: [ // these are all currently Kubernetes-centric, will need to update for serverless
+      {
+        kind: 'kubernetesNamespace'
+        namespace: 'myapp'
+      }
+      {
+        kind: 'kubernetesMetadata'
+        labels: {
+          'team.contact.name': 'frontend'
+        }
+      }
++     // add other serverless extensions as applicable
+    ]
+  }
+}
+```
+
+### Step 3: Define a Radius Container for serverless
+
+The user defines a Radius Container within the application definition (e.g. `app.bicep`) including all the platform-agnostic [`container` properties](https://docs.radapp.io/reference/resource-schema/core-schema/container-schema/#container), e.g. `image`, `env`, etc. and deploy the application container to the serverless environment. The user may also set platform specific configurations via the `extensions` or `runtimes` properties, but any properties that are not applicable to the targeted compute platform will be ignored at deploy time. Note that the container definitions should be container runtime platform-agnostic, thus this same container definition can be deployed to both Kubernetes and serverless compute platforms.
+
+```diff
+resource demo 'Applications.Core/containers@2023-10-01-preview' = {
+  name: 'demo'
+  properties: {
+    application: application
+    container: {
+      image: 'ghcr.io/radius-project/samples/demo:latest'
+      ports: {
+        web: {
+          containerPort: 3000
+        }
+      }
++     // all the other container properties should also be implemented (e.g. `env`, `readinessProbe`, `livenessProbe`, etc.)
+    }
+    extensions: [
++     {
```

Comment:
Would we need to specify a runtime under `extensions`, or are we planning to match element names? If we are matching element names, we could run into naming conflicts where the same name means different things on different platforms.

Reply:
I think as a part of the tech design we should address this and change the schema if it makes more sense because I agree with your point.